File: python/clock/clock.py
MINS_PER_HOUR = 60
MINS_PER_DAY = 24 * MINS_PER_HOUR


class Clock:
    def __init__(self, hour, minute):
        self.minutes = (hour * MINS_PER_HOUR + minute) % MINS_PER_DAY

    def __repr__(self):
        return f"{self.minutes // MINS_PER_HOUR:02}:{self.minutes % MINS_PER_HOUR:02}"

    def __eq__(self, other):
        return self.minutes == other.minutes

    def __add__(self, minutes):
        return Clock(0, self.minutes + minutes)

    def __sub__(self, minutes):
        return Clock(0, self.minutes - minutes)

File: ruby/clock/clock.rb
class Clock
  MINUTES_PER_HOUR = 60
  MINUTES_PER_DAY = MINUTES_PER_HOUR * 24

  attr_reader :minutes

  def initialize(hour: 0, minute: 0)
    self.minutes = minute + hour * MINUTES_PER_HOUR
  end

  def minutes=(minutes)
    @minutes = ((minutes % MINUTES_PER_DAY) + MINUTES_PER_DAY) % MINUTES_PER_DAY
  end

  def +(other)
    Clock.new(hour: 0, minute: minutes + other.minutes)
  end

  def -(other)
    Clock.new(hour: 0, minute: minutes - other.minutes)
  end

  def eql?(other)
    hash == other.hash
  end

  def hash
    minutes.hash
  end

  def ==(other)
    minutes == other.minutes
  end

  def to_s
    format('%02d:%02d', *minutes.divmod(MINUTES_PER_HOUR))
  end
end
